### DISCUSSION TITLE
credit card number format

### DISCUSSION THREAD

**Discussion by pedroR** [react-hook-form/react-hook-form#8964] [Q&A] [answered]
Hi, I am trying to format a credit card number. I want to add a space after each 4 character. I can see the correct formatted value on a console.log ( commented bellow), but it doesn't show properly in the component. I may be doing something wrong on passing the info to the component.

FYI I am using my own input component.  `<TextInput>`

I have the following code: 

```
interface CreditCard {
  nameOnCard: string;
  cardNumber: number;
  validity:string;
  cvv: number
}
....
```

```
const [cardNumberInput, setCardNumberInput] = useState<string>();

const {
    handleSubmit,
    register,
    reset,
    watch,
    control,
    formState: { isDirty, isValid },
  } = useForm<CreditCard>({
    mode: "onChange",
  });

const formatCardNumber = (value: string) => {
    const val = value
        .replace(/\s|[^0-9]+/g, "")
        .match(/.{1,4}/g)
        ?.join(" ") ?? "";

    console.log("-->>>  ", val); // returns the correct value with a space at each 4 numbers
    setCardNumberInput(val);  // set the correct value to the state variable

....

return (
...

<Controller
          control={control}
          name={"cardNumber"}
          render={({ field: { onChange } }) => (
            <TextInput
              name={"cardNumber"}
              title={"Card Number"}
              // inputValue={watch("cardNumber")}
              inputValue={cardNumberInput}
              register={register}
              onTextChange={(value: string) => formatCardNumber(value)}
              pattern={"[\d ] {16}"}
              mandatory
            />
          )}
        />

...
)
```

What do I need to fix to add a space after each 4 digit?

**Comment by Moshyfawn** [comment]
Please, take some time to create a minimal codesandbox reproduction

**Reply by pedroR** [reply]
Hi @Moshyfawn 
I did a minimalist codesandbox to reproduce my doubt 
You can take a look: https://codesandbox.io/s/sleepy-snow-bxinz9

Thanks

**Reply by Moshyfawn** [reply]
Awesome, thx, will take a look once available

**Comment by Moshyfawn** [comment, marked as answer]
First of all, your underlining input's using the `register` API, which is only helpful if you don't need to manipulate the input value, to use the input in **uncontrolled** way. Whenever your task is to **format a value**, **update before change** - you need the `Controller` component / `useController` hook, to use your input in the **controlled** manner.

Another issue is, you are trying to use the `Controller` component **with** the `register` API, which is a no-go; it's **either or**.

I haven't gone deep to learn how your your `Textinput` component works, but I see that the register method is an optional prop, so here what I'd suggest:
- To avoid passing the `register` method down to your `TextInput`
- To pass the `Controller` value into your `TextInput` as `inputValue`

Check out [this working csb](https://codesandbox.io/s/aged-shape-wjj9m3?file=%2Fsrc%2Ftext-input%2FTextInput.tsx%3A1933-1951)

I'd also recommend to clean up your `TextInput`; you have a lot of different props from both controlled and uncontrolled approaches

**Reply by pedroR** [reply]
It works great. I will take in consideration your notes. Thanks.

**Reply by Moshyfawn** [reply]
No worries, glad I could help!
